h/util/session_tracker.py

```diff
@@ -0,0 +1,74 @@
+from copy import copy
+from enum import Enum
+
+from sqlalchemy.event import listen
+from sqlalchemy.orm.util import identity_key
+
+
+class ObjectState(Enum):
+    ADDED = 'added'
+    DELETED = 'deleted'
+    CHANGED = 'changed'
+
+
+class Tracker(object):
+    """
+    Observer which tracks whether a SQLAlchemy `Session` has uncommitted changes.
+    """
+
+    def __init__(self, session):
+        self.session = session
+
+        # Changes which have been flushed to the database (and thus not present
+        # in `session.new`, `session.dirty` or `session.deleted) but have not
+        # yet been _committed_.
+        self._flushed_changes = {}
+
+        listen(session, 'after_flush', self._after_flush)
+        listen(session, 'after_commit', self._after_commit)
+        listen(session, 'after_rollback', self._after_rollback)
+
+    def uncommitted_changes(self):
+        """
+        Return a list of changes to the `Session` which have not yet been
+        _committed_ to the DB.
+
+        The result is a list of (model class, primary key, ObjectState) tuples.
```

Comment:
Am I right in thinking this comment is now out of date, since 855daa89d2f472d4e54e2c8488f8e1e662df1976 changed the return type to an `(identifier, state)` tuple?

(Tangential note: I see in the Pyramid docs that the `identity_key` tuple [gets an extra element in version 1.2](http://docs.sqlalchemy.org/en/latest/changelog/migration_12.html#change-4137), so let’s make sure we don’t end up relying on the structure of the key if we can avoid it.)

Reply:
Yes indeed. The result of `identity_key` is treated as an opaque value here and in the tests so I think we should be OK wrt. changes in future SQLAlchemy releases.